include/inexor/vulkan-renderer/camera.hpp

```diff
@@ -1,127 +1,216 @@
-    /// @note The view matrix should only be updated if necessary.
-    void update_view_matrix();
+namespace directions {
+/// The default value of the Camera's front vector
+static constexpr glm::vec3 DEFAULT_FRONT{0.0f, 0.0f, 1.0f};
+/// The default value of the Camera's right vector.
+static constexpr glm::vec3 DEFAULT_RIGHT{1.0f, 0.0f, 0.0f};
+/// The default value of the Camera's up vector.
+static constexpr glm::vec3 DEFAULT_UP{0.0f, 1.0f, 0.0f};
+}; // namespace directions
 
-    CameraType m_type{CameraType::LOOKAT};
+enum class CameraMovement { FORWARD, BACKWARD, LEFT, RIGHT };
 
-    glm::vec3 m_position{};
-    glm::vec3 m_rotation{};
+// TODO: Implement more camera types.
+enum class CameraType { CAMERA_TYPE_LOOKAT };
 
-    float m_movement_speed{1.0f};
+/// @warning Not thread safe!
+class Camera {
+private:
+    /// The type of the camera. Currently only one type is implemented.
+    CameraType m_type{CameraType::CAMERA_TYPE_LOOKAT};
+    /// The start position of the camera.
+    glm::vec3 m_position{0.0f, 0.0f, 0.0f};
+    /// The vector of direction in which the camera is looking.
+    glm::vec3 m_front{DEFAULT_FRONT};
+    /// The vector of direction which points to the right.
+    glm::vec3 m_right{DEFAULT_RIGHT};
+    /// The vector which indicates "upwards".
+    glm::vec3 m_up{DEFAULT_UP};
+    /// The world vector which indicates "upwards".
+    glm::vec3 m_world_up{0.0f, 1.0f, 0.0f};
+    glm::mat4 m_view_matrix{};
+    glm::mat4 m_perspective_matrix{};
+
+    /// The camera's yaw angle.
+    float m_yaw{0.0f};
+    /// The camera's roll angle.
+    float m_roll{0.0f};
+    /// The camera's pitch angle.
+    float m_pitch{0.0f};
+    /// The camera's minimum pitch angle.
+    /// Looking straight downwards is the maximum pitch angle.
+    float m_pitch_min{-89.0f};
+    /// The camera's maximum pitch angle.
+    /// Looking straight upwards is the maximum pitch angle.
+    float m_pitch_max{+89.0f};
+    /// The camera's field of view.
+    float m_fov{90.0f};
+    /// The camera's maximum field of view.
+    float m_fov_max{90.0f};
+    /// The camera's minimum field of view.
+    float m_fov_min{20.0f};
+    /// The zoom step when zooming in or out.
+    float m_zoom_step{10.0f};
+    /// The camera's rotation speed.
     float m_rotation_speed{1.0f};
+    /// The camera's movement speed.
+    float m_movement_speed{2.0f};
+    /// The camera's aspect ratio (width divided by height).
+    float m_aspect_ratio{1920.0f / 1080.0f};
+    /// The sensivity of the mouse.
+    float m_mouse_sensitivity{0.005f};
+    /// The camera's near plane.
+    float m_near_plane{0.001f};
+    /// The camera's far plane.
+    float m_far_plane{1000.0f};
 
-    float m_fov{0.0f};
-    float m_z_near{0.0f};
-    float m_z_far{0.0f};
+    /// The keys for the movement FORWARD, BACKWARD, LEFT, RIGHT.
+    std::array<bool, 4> m_keys{false, false, false, false};
+    /// Signals if the matrices need to be recalculated
```

Comment:
Signals is a bad word here as it has nothing to do with actual signals.

Reply:
I thought so as well.

Reply:
Done.